lib/Conversion/TorchToLinalg/Reduction.cpp

```diff
@@ -244,67 +411,39 @@ class ConvertReductionOp : public ConversionPattern {
   matchAndRewrite(Operation *op, ArrayRef<Value> operands,
                   ConversionPatternRewriter &rewriter) const override {
     if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
-      return failure();
-
-    // Every reduce operation must set a value for the `dimSet`,
-    // `tensorOperand`, and `keepDim` in accordance with their specification.
-    DenseSet<int64_t> dimSet;
-    Value tensorOperand;
-    bool keepDim = false;
-    if (isa<AtenSumOp>(op) || isa<AtenMaxOp>(op)) {
-      tensorOperand = operands[0];
-      auto inputType = tensorOperand.getType().cast<RankedTensorType>();
+      return rewriter.notifyMatchFailure(
+          op, "invalid operand or result types to use with linalg on tensors");
 
-      // `AtenSumOp` and `AtenMaxOp` reduces along all the dimensions of the
-      // input tensor.
-      for (int64_t i = 0; i < inputType.getRank(); i++)
-        dimSet.insert(i);
-    } else if (auto sumDimIntListOp = dyn_cast<AtenSumDimIntListOp>(op)) {
-      tensorOperand = operands[0];
-      auto inputType = tensorOperand.getType().cast<RankedTensorType>();
-
-      if (!matchPattern(sumDimIntListOp.keepdim(),
-                        m_TorchConstantBool(&keepDim)))
-        return failure();
-
-      SmallVector<int64_t> dimList;
-      if (!matchPattern(sumDimIntListOp.dim(), m_TorchConstantIntList(dimList)))
-        return failure();
-      for (auto dim : dimList) {
-        // Torch allows for negative values in dimSet to go in reverse
-        // order in the dimensions of the input tensor.
-        dim = dim >= 0 ? dim : dim + inputType.getRank();
-        // Drop invalid dimensions
-        if (dim < inputType.getRank())
-          dimSet.insert(dim);
-      }
-    } else {
-      return rewriter.notifyMatchFailure(op, "not a supported reduce op");
-    }
+    auto opInfo = computeReductionOpInfo(op, operands, rewriter);
```

Comment:
I noticed that `auto` is used a lot in these changes. Here we follow the LLVM standard of using `auto` only when it improves readability (https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable). This line is an example of a case where knowing the type would help the reader understand the body better without having to look up the signature of `computeReductionOpInfo`.

Can you go over the `auto`s in these changes and add types back in places where it is not clear from context what the type should be?

Reply:
Makes sense.  Fixed in commit c49cf368.  I've changed the code that I added to use types wherever it isn't clear, but I've left the unchanged lines as they are.